### ISSUE TITLE
Revisit command-buffer simultaneous-use

### ISSUE THREAD

**Issue by EwanC** [KhronosGroup/OpenCL-Docs#891]
During the course of writing CTS test for cl_khr_command_buffer, in particular https://github.com/KhronosGroup/OpenCL-CTS/pull/1584, questions have arisen about the behaviour of [CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#_modifications_to_section_4_2_of_the_opencl_api_specification) and how usable it is for developers.

E.g - Do the following examples exhibit simultaneous-use behaviour?

```cpp
// Case 1: queue is an out-of-order queue.
// Note: No event dependencies!
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
clFinish(queue);

// Case 2: queue0 and queue1 are in-order queues.
clEnqueueCommandBufferKHR(1, &queue0, cmdbuf, 0, NULL, NULL);
clEnqueueCommandBufferKHR(1, &queue1, cmdbuf, 0, NULL, NULL);
clFlush(queue0); // these flushes may not be required
clFlush(queue1);
clFinish(queue0);
clFinish(queue1);
```

In my mind, a lot of the behaviour of simultaneous-use actually depends on whether a vendor implements queues eagerly or lazily. In lazy execution (what happens in Codeplay' implementation for example) we wouldn't start submitting anything until the clFinish()/cFlush(), so the pending count keeps on increasing until then, even for in-order queues. In an eager queue implementation, the first enqueue would be run straight away, and may indeed have finished by the time the host code calls the second command buffer.

Depending on an unreported implementation detail is hard for a user developing code to reason about
though, so maybe we can do better here. Before the cl_khr_command_buffer extension becomes final from provisional we should confirm if the currently specified behaviour is what we want, or whether we should revisit this property.

**Comment by bashbaug** [comment]
I think this is another case to consider, especially for implementations that implement queues lazily (AKA "batched").  I think this would be considered simultaneous use?

```c++
// Case A: queue is an in-order queue.
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
clFinish(queue);
```

The in-order queue semantics guarantee that the first command buffer execution will "complete" before the second command buffer execution "starts", so they will never be _executing_ "simultaneously", but the second command buffer execution may certainly be "queued" or "submitted" before the first command buffer execution is "complete".

**Comment by EwanC** [comment]
I've been mulling this over in the background, and these are my thoughts so far on ways forward. Nothing concrete but posting to keep the discussion ticking over and maybe help flush out more ideas:

1) Change functionality from optional to mandatory. Would be the most beneficial choice from the user perspective, but increases implementation burden. Possible implementations for simultaneous-use that don't have native support could be to A) deep-copy the command-buffer and enqueue the copy, B) runtime schedules the simulatenous submission after the first submission by creating an implicit dependency, C) runtime blocks on simultaneous submission in command-buffer enqueue entry-point until the first submission is complete and flushes the queue (effectively the runtime does the `clFinish(queue)` rather than application code)

2) Change the definition of simultaneous execution from a course "if pending count > 1" to something more granular that takes into account dependencies. For example, "if pending count > 1 and the command-buffer being enqueued has a no dependencies on pending executions of the command-buffer". The ordering dependency means no simulatneos execution of command-buffer could occur, so I think Ben's in-order snippet would be valid on a runtime without simultaneous-use support.
    Cross queue & out-of-order queue submissions of the same command-buffer would still need user invocations of blocking waits if there is no explicit event dependency. If we wanted to avoid that, I think we're back at 1B) above.

What i'm wary of with suggestion 2) is that the implementation effort reqired for vendors that don't support simultaneous enqueue, we'd be as well specifying 1) to cover more use-cases.

**Comment by EwanC** [comment]
> Interesting... I think this implies that all three of the cases above should be considered "simultaneous use", agreed?
>
> Basically, to use a command-buffer "non-simultaneously" you need to observe that the command-buffer is no longer in the pending state, which could be done directly (by querying the command-buffer state) or perhaps done indirectly (say, by calling clFinish on the command-queue executing the command-buffer?)


Yup, I think we're aligned on the interpretation. So I don't think we need any functional changes to the spec, and the slightly awkward code in the CTS that motivated this discussion is a result of the need to observe the completion of the previous enqueue.

We could make improvements in the description of simultaneous-use to better clarify these findings though, that 1) simultaneous-use isn't just about when the command-buffer is executed, but when it's enqueued. 2) The onus is on the user to provide explicit synchronization in the non-simultaneous case to guarantee that the previous enqueue has completed.

> IMHO it's an interesting thought experiment to consider if we could write some sort of validation layer to check if an application is properly using command buffers that do not support simultaneous use. 

I think that should be possible, there might be other transitive cases like a blocking command, e.g clEnqueueReadBuffer, on the same queue with a previous pending command-buffer enqueue dependency. But it does sound possible, and useful!

> For fun, do we agree that this would technically be considered simultaneous use?

Yup, I don't think it would be guaranteed by the spec that the case isn't simultaneous use.

**Comment by SunSerega** [comment]
> If an application calls `clWaitForEvents` to wait on the event associated with the command buffer execution, remove the command-buffer from the possibly pending list.

What about `clSetEventCallback`? For instance, it could be said:

- While the callback is executing, a new enqueue of the command buffer is allowed on the callback thread. But after the callback finishes, a call to `clWaitForEvents` is once again needed to ensure no simultaneous use. Even if the callback was triggered by `clSetUserEventStatus` and immediately executed on the same thread.

This might be overly limiting. But I think the alternative is to always consider new enqueue a non-simultaneous use if `clSetEventCallback` was ever called on any related event.

And anyway, while the limitation to blocking calls in the callback may make it harder to work around, I believe it is better to enqueue everything new in the callback and only send blocking calls like `clFinish` (which are not checked for simultaneous use like the enqueue) to be executed in a separate thread.

**Comment by EwanC** [comment]
We've been comparing the simultaneous use semantics to Vulkan, but all the other APIs I've looked at don't provide this functionality (CUDA-Graph, HIP-Graph, Level-Zero, DX12) and limit a single instance of a command-buffer to be executing at once.
However, the majority of these API provide a clone API (couldn't see on in DX12) which could be used to support the use-case of submit->update->submit without blocking on the host. As a user could create multiple clones in a warm up stage of the application, then in the hot path: submit the original command-buffer, update a clone, submit that clone while the original command-buffer is still executing.

Although creating these clones might be more work for the user, semantics across implementations would be clearer. One concern i'm not sure about is if cloning would be extra overhead for hardware which driver's do natively support simultaneous use. However we could maybe define the clone semantics in such a way that the clone isn't a complete deep copy. For example, in 
the [Level Zero clone API](https://spec.oneapi.io/level-zero/latest/core/api.html#zecommandlistcreatecloneexp) there's the wording "The cloned command list must be destroyed prior to the source command list." which means that some resources wouldn't need to be copied over to the clone.

In summary, I'm floating the idea that we could introduce a command-buffer clone API to replace simultaneous-use.

**Comment by bashbaug** [comment]
Noting for completeness: there was also some discussion of a clone API in internal issue 251, in two places actually.

I agree this is something we should consider.

**Comment by FreddieWitherden** [comment]
> We've been comparing the simultaneous use semantics to Vulkan, but all the other APIs I've looked at don't provide this functionality (CUDA-Graph, HIP-Graph, Level-Zero, DX12) and limit a single instance of a command-buffer to be executing at once. However, the majority of these API provide a clone API (couldn't see on in DX12) which could be used to support the use-case of submit->update->submit without blocking on the host. As a user could create multiple clones in a warm up stage of the application, then in the hot path: submit the original command-buffer, update a clone, submit that clone while the original command-buffer is still executing.
> 
> In summary, I'm floating the idea that we could introduce a command-buffer clone API to replace simultaneous-use.

CUDA and HIP go further by providing implicit cloning.  That is to say I can submit a graph, then update a node (say by `cuGraphExecKernelNodeSetParams`), submit it again, and everything is fine in the sense that the first queuing will use the old parameters and the second queuing will use the new parameters.  This is because "The modifications only affect future launches of hGraphExec. Already enqueued or running launches of hGraphExec are not affected by this call. hNode is also not modified by this call."  Broadly, this is what developers want.  Managing what is queued, what is executing, and what has executed is the runtimes problem and it is in a better place to know how to handle such modifications (as in knowing if to make a clone or not).

**Comment by EwanC** [comment]
Thanks for the input @FreddieWitherden We've got wording in `clUpdateMutableCommandsKHR` already that it only affects future enqueues of a command-buffer, which I hoped would imply "already enqueued or running launches of the command-buffer are not affected by this call" but maybe should say that explicitly too.

I think an interesting difference to CUDA is the wording in [cudaGraphLaunch](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g1920584881db959c8c74130d79019b73) about ordering with previous executable graph launches.

> Only one instance of graphExec may be executing at a time. Each launch is ordered behind both any previous work in stream and any previous launches of graphExec. To execute a graph concurrently, it must be instantiated multiple times into multiple executable graphs. 

Creating implicit dependencies between executions of the same command-buffer isn't something we've talked about before, but if I'm reading that correctly is what CUDA is doing for executable graphs. So I think one difference with the CUDA model v explicit clone, is that with the CUDA model a submit->update->submit flow would prevent submit 1 and submit 2 from overlapping execution. Whereas if the user did an explicit clone and submitted that for the second commit, it could run with the first submit.

---
Do you have a link to your wrapper @SunSerega? It sounds interesting, and i'm not quite following yet how the subgraph/payload concepts could be applied to command-buffers other than the idea of creating some handle to the graph inputs/outputs in advance.

**Comment by FreddieWitherden** [comment]
> Creating implicit dependencies between executions of the same command-buffer isn't something we've talked about before, but if I'm reading that correctly is what CUDA is doing for executable graphs. So I think one difference with the CUDA model v explicit clone, is that with the CUDA model a submit->update->submit flow would prevent submit 1 and submit 2 from overlapping execution. Whereas if the user did an explicit clone and submitted that for the second commit, it could run with the first submit.

So it is important to remember in CUDA/HIP there is a graph and then there are *instances* of the graph.  The instances are created with `cuGraphInstantiateWithFlags` and can then have their arguments tweaked (to an extent).  Only instances of a graph can be launched, and a single instance will only ever run once.  So if you want parallel execution you would instantiate multiple instances.  (As for the why: I suspect instantiating a graph creates a bunch of CUDA streams under the hood; submitting a graph to a stream actually results in the kernels being submitted to these graph-private streams and only a synchronization marker is actually added to the stream you submitted the graph too.)

**Comment by SunSerega** [comment]
> Do you have a link to your wrapper @SunSerega? It sounds interesting, and i'm not quite following yet how the subgraph/payload concepts could be applied to command-buffers other than the idea of creating some handle to the graph inputs/outputs in advance.

<details>
<summary>About my wrapper</summary>

https://github.com/SunSerega/POCGL
In this case, I was talking about this in particular:
https://github.com/SunSerega/POCGL/blob/unstable/Modules.Packed/OpenCLABC.pas

```
## uses OpenCLABC;

// "A" is an array, backed by GPU memory
// "ind" is a sub-graph, returning an integer value
// "Result" is a CommandQueue returning nil aka. nothing
function MakeQueue(A: CLArray<byte>; ind: CommandQueue<integer>): CommandQueueNil;
begin
  // HPQ = HostProcedureQueue. A command wrapper for a CPU code returning nothing
  // In this case the first action is to create a new thread and make it sleep for a second
  // Pretty ugly way to add a delay, but just an example
  Result := HPQ(()->Sleep(1000), need_own_thread := true);
  // Addition means sequential execution
  // Create CommandContainerQueue with a single buffer write
  Result += A.MakeCCQ.ThenWriteValue(1, ind).DiscardResult;
  // Basically the result is a write command, that first waits for a second before executing
end;

var A := new CLArray<byte>(4);
var Q := CQNil; // an instance of a ConstQueueNil singleton
var P := new ParameterQueue<integer>('par name for debug');

// Multiplication means parallel execution
// (doesn't make much sense to write to the same buffer in parallel tho...)

// "ind" is a ConstQueue
Q *= MakeQueue(A, CQ(0));
// Can also convert to ConstQueue implicitly
Q *= MakeQueue(A, 1);

// "ind" is a ParameterQueue
Q *= MakeQueue(A, P);

// HFQ is like HPQ but with a return value
// "ind" is of a special internal type,
// neither const nor parameter. basically dynamic
Q *= MakeQueue(A, HFQ(()->3));

CLContext.Default.SyncInvoke(
  Q + A.MakeCCQ.ThenGetArray.ThenUse(a->Println(a)),
  P.NewSetter(2)
); // Prints [1,1,1,1]
```

---

</details>

As for how it might relate to command buffers here... Well, not directly.
I mainly set it as an example of how fully immutable command graphs can exist. I'm not sure if command buffers can change enough at this point.

The issue header still says command buffers are provisional, but even the simplest implementation of this idea would still change a lot to work... For instance:
```
cl_command_input inp_const = clCreateCommandInputConst(
	0, nullptr, // command buffers to execute as payload before this input becomes valid
	sizeof(int), 123, // the const return value
	ec // error code return
);

cl_command_input inp_par = clCreateCommandInputParameter(
	0, nullptr,
	sizeof(int), // the size of parameter value
	ec
);

void make_dynamic_result(void* ptr) {
	int* res = ptr;
	*res = 3;
}
...
cl_command_input inp_dyn = clCreateCommandInputFunction(
	0, nullptr,
	sizeof(int), make_dynamic_result, // the callback to query value after all the command buffers
	ec
);

// (use these three as inputs for special commands in the command buffer)

int par_val = 2;
cl_command_buffer_parameter_value par_values[] = {
	{ inp_par, sizeof(par_val), &par_val }, // specify info about parameter size again, as a sanity check
};

ec = clEnqueueCommandBufferParameterisedKHR(
	//TODO All the normal parameters
	1, par_values
);
```

**Comment by EwanC** [comment]
This topic came up again in the teleconference a couple of weeks ago in the context of [SYCL-Graph](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc) which has the wording:

> If graph is submitted multiple times, dependencies are automatically added by the runtime to prevent concurrent
> executions of an identical graph.

This would correspond to the following SYCL-RT implementation on an OpenCL backend if the OpenCL device supports simultaneous-use: 

**Psuedocode A**
```
cl_event e;
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, &e); // First submission of a graph to a sycl queue
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 1, &e, NULL); // second submission of graph to same sycl queue
```

However if the device doesn't support simultaneous-use, a host-side wait is required. As the value of the command-buffer pending count is unknown the time of its second enqueue, making it a possibility that an error could be thrown if the device doesn't support simultaneous-use:

**Psuedocode B**
```
cl_event e1;
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, &e1);
clWaitForEvents(1, &e1)
cl_event e2;
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, &e2); // e2 used as a dependency for any future enqueues.
```

Replacing simultaneous-use in cl_khr_command_buffer with the [clone solution previously mentioned in the thread](https://github.com/KhronosGroup/OpenCL-Docs/issues/891#issuecomment-1935706858) would allow implementing the SYCL semantics using *Psuedocode A* without needing to consider optional OpenCL device capabilities.

**Comment by bashbaug** [comment]
We had a long discussion about this issue in the December 3rd teleconference.  After discussion, some of the things we would like to accomplish are:

1. Clarify the definition of simultaneous use, so it is precisely defined, and we can go through the examples in this issue and answer whether each scenario is or is not simultaneous use.
2. Ensure that the definition of simultaneous use makes sense to developers, and ideally, that developers get an informative error message if they erroneously write code that requires simultaneous use with an implementation that does not support simultaneous use.
3. Understand what higher-level language runtimes (such as SYCL) will need to do to write correct code on devices with and without simultaneous use.
4. Understand how we can test all of this.

As a next step, for (1), we are going to revisit the command buffer state diagram to clarify the state transitions, and especially the state transition from “pending” back to “executable”.

**Comment by bashbaug** [comment]
Discussed additionally in the December 10th teleconference.  Here are a few other things to add to the list we would like to accomplish:

5. Support additional scenarios safely even on systems that do not support simultaneous use, specifically back-to-back enqueues without overlapping execution, say in an in-order queue.
6. Clarify the interaction with mutable dispatch.
7. Consider being more test driven, by writing tests to see how implementations behave before working on the exact spec text.

**Comment by EwanC** [comment]
> As a next step, for (1), we are going to revisit the command-buffer state diagram to clarify the state transitions, and
especially the state transition from “pending” back to “executable”.

Looked at what the [Vulkan state diagram](https://registry.khronos.org/vulkan/specs/latest/html/vkspec.html#commandbuffers-lifecycle) for this state transition says, and I think the relevant part is "A [synchronization command](https://registry.khronos.org/vulkan/specs/latest/html/vkspec.html#synchronization) should be used to detect when this occurs."

Which if we exported that wording for OpenCL I think would mean the spec asking users to do the following for non-simultaneous-use code:
```diff
// Case A: queue is an in-order queue.
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
+ // user needs use an event here from last enqueue, or do `clFinish(queue)`.
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
clFinish(queue);
```

Which is definitely clearer messaging to say that any two enqueues of the same command-buffer are simultaneous use unless there is a blocking host wait between them, but a relatively awkward change to ask developers to make. Especially if for (5) we don't want a user to have to insert any new code between the 2 enqueue calls for manual checking of state.

**Comment by joshqti** [comment]
I think the OpenCL command buffer state diagram is clear, a non-simulteneous command buffer may not be enqueued at all if the command buffer is pending, which corresponds to the last enqueue being CL_COMPLETE.  I believe this makes sense. 

An event dependency is not sufficient to prevent non-simultaneous use.  As soon as two commands enter the queued state, they may start modifying the state of the command buffer, for example the queued command may modify the state while a command is running.  My interpretation of non-simultaneous use is implementations do not need to handle these sort of situations, with multi-buffering or some other scheme.

**Comment by EwanC** [comment]
Reading back over this, I think this bullet on our list of what we'd like to achieve is the crux point:

> 5. Support additional scenarios safely even on systems that do not support simultaneous use, specifically back-to-back enqueues without overlapping execution, say in an in-order queue.

Without this requirement I think we could make our definition more understandable/testable with wording like (definitely draft wording):

> "Simultaneous use occurs when a command-buffer is submitted or updated without prior blocking host synchronization on the completion of any previous submissions of the command-buffer."

If we want to support back-to-back in-order queue submissions on a device without simultaneous-use, without the user having to insert any extra code between those enqueue statements. Then I think the gap between API usage with and without simultaneous-use become very close and any definition might be too nuanced for having simultaneous-use as a concept to be the best way forward. 

This path forward might make vendor agnostic user code more verbose by introducing an idiom like below. But at least it's a clear message of what simultaneous-use is, and users might put pressure on vendors to support the capability to avoid it.

```cpp
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
if (!simultaneous_use_support) {
 clFinish(queue);
}
clEnqueueCommandBufferKHR(1, &queue, cmdbuf, 0, NULL, NULL);
```

**Comment by aharon-abramson** [comment]
I think that the primary reason behind non-simultaneous use is to make sure the command buffer has all the resources it needs when it starts executing. If we allow updating it while it has been enqueued and not completed (like in the example above), it means that either a command's parameters or arguments are modified while it might be running, or it requires the command buffer to allocate new sources dynamically and update them.
Since neither option seems desirable, I believe we should disallow it.

**Comment by EwanC** [comment]
From 17/06/2025 teleconference, we said we'd first try to defined simultaneous-execution to look at the error behavior from a different angle.

I've had a go a doing this using the "prerequisite" language from the [execution model](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_execution_model). Where I'm assuming the term "prerequisite" covers ancestors of direct dependencies in the command DAG. i.e. For A->B->C, then A is a prerequisite of C, as it's a dependency of the direct dependency B.

My attempt at defining this is

> Simultaneous Execution is when a command-buffer is submitted for execution without a prerequisite on previous submissions of the same command-buffer which are not in the Complete state.


```
  clEnqueueCommandBufferKHR(1, &out_of_order_queue, cmdbuf, 0, NULL, NULL);  // A
  if (device_does_not_support_sim_execution) {
    // User needs a blocking wait to guarantee command A has completed before submitting command B
    clFinish(out_of_order_queue); 
  }
  clEnqueueCommandBufferKHR(1, &out_of_order_queue, cmdbuf, 0, NULL, NULL);  // B
```

**Comment by bashbaug** [comment]
Mentioning for completeness, if we adopt this wording:

> Simultaneous Execution is when a command-buffer is submitted for execution without a prerequisite on all the previous submissions of the same command-buffer which are not in the Complete state.

Then the snip above with `clFinish` is one of the ways to handle the case without simultaneous execution, but not the only one.  Specifically, I think all of these sequences would work, also:

Command-queue barrier dependency between A and B:

```c++
  clEnqueueCommandBufferKHR(1, &out_of_order_queue, cmdbuf, 0, NULL, NULL);  // A
  clEnqueueBarrierWithWaitList(out_of_order_queue, 0, nullptr, nullptr);
  clEnqueueCommandBufferKHR(1, &out_of_order_queue, cmdbuf, 0, NULL, NULL);  // B
```

Explicit event dependency:

```c++
  clEnqueueCommandBufferKHR(1, &out_of_order_queue, cmdbuf, 0, NULL, &eventA);  // A
  clEnqueueCommandBufferKHR(1, &out_of_order_queue, cmdbuf, 1, &eventA, NULL);  // B
```

Implicit in-order queue dependency:

```c++
  clEnqueueCommandBufferKHR(1, &in_order_queue, cmdbuf, 0, NULL, NULL);  // A
  clEnqueueCommandBufferKHR(1, &in_order_queue, cmdbuf, 0, NULL, NULL);  // B
```
